store/instance.go

```diff
@@ -759,3 +759,123 @@ func findInstanceQuery(find *api.InstanceFind) (string, []interface{}) {
 
 	return strings.Join(where, " AND "), args
 }
+
+// InstanceMessage is the mssage for instance.
+type InstanceMessage struct {
+	InstanceID   string
+	Title        string
+	Engine       db.Type
+	ExternalLink string
+	Deleted      bool
+	DataSources  []*DataSourceMessage
+}
+
+// GetInstanceV2 gets an instance by the resource_id.
+func (s *Store) GetInstanceV2(ctx context.Context, resourceID string) (*InstanceMessage, error) {
+	tx, err := s.db.BeginTx(ctx, nil)
+	if err != nil {
+		return nil, FormatError(err)
+	}
+	defer tx.Rollback()
+
+	var instanceMessage InstanceMessage
+	var rowStatus string
+	var instanceID int
+	if err := tx.QueryRowContext(ctx, `
+		SELECT
+			id,
+			resource_id,
+			name,
+			engine,
+			external_link,
+			row_status
+		FROM instance
+		WHERE resource_id = $1`,
+		resourceID,
+	).Scan(
+		&instanceID,
+		&instanceMessage.InstanceID,
+		&instanceMessage.Title,
+		&instanceMessage.Engine,
+		&instanceMessage.ExternalLink,
+		&rowStatus,
+	); err != nil {
+		if err == sql.ErrNoRows {
+			return nil, nil
+		}
+		return nil, FormatError(err)
+	}
+	instanceMessage.Deleted = convertRowStatusToDeleted(rowStatus)
+
+	dataSourceList, err := s.listDataSourceV2(ctx, instanceID)
+	if err != nil {
+		return nil, FormatError(err)
+	}
+	instanceMessage.DataSources = dataSourceList
+
+	if err := tx.Commit(); err != nil {
+		return nil, FormatError(err)
+	}
+
+	return &instanceMessage, nil
+}
+
+// ListInstanceV2 lists all instance.
+func (s *Store) ListInstanceV2(ctx context.Context, showDeleted bool) ([]*InstanceMessage, error) {
+	where, args := []string{"1 = 1"}, []interface{}{}
+	if !showDeleted {
+		where, args = append(where, fmt.Sprintf("row_status = $%d", len(args)+1)), append(args, api.Normal)
+	}
+
+	tx, err := s.db.BeginTx(ctx, nil)
+	if err != nil {
+		return nil, FormatError(err)
+	}
+	defer tx.Rollback()
+
+	var instanceMessages []*InstanceMessage
+	rows, err := tx.QueryContext(ctx, `
+		SELECT
+			id,
+			resource_id,
+			name,
+			engine,
+			external_link,
+			row_status
+		FROM instance
+		WHERE `+strings.Join(where, " AND "),
+		args...,
+	)
+	if err != nil {
+		return nil, FormatError(err)
+	}
+	defer rows.Close()
+	for rows.Next() {
+		var instanceMessage InstanceMessage
+		var rowStatus string
+		var instanceID int
+		if err := rows.Scan(
+			&instanceID,
+			&instanceMessage.InstanceID,
+			&instanceMessage.Title,
+			&instanceMessage.Engine,
+			&instanceMessage.ExternalLink,
+			&rowStatus,
+		); err != nil {
+			return nil, FormatError(err)
+		}
+		instanceMessage.Deleted = convertRowStatusToDeleted(rowStatus)
+		dataSourceList, err := s.listDataSourceV2(ctx, instanceID)
```

Comment:
The host/port/etc in the instance will be added to data source list as ADMIN data source. Or was it also added to data source table?

Reply:
How about add them into the data source table?

Reply:
Yeah, we can address this later.